### PR TITLE
lnwallet: fix closechannel for P2TR external addr

### PR DESCRIPTION
**Pull request by starius**
## Change Description

Customize the itest with the type of external delivery address. Test with P2TR address type in addition to P2WKPH.

## Steps to Test

```
make itest icase='coop close with external delivery'
```

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: lnwallet/interface.go [hunk 1]
```diff
@@ -636,6 +636,16 @@ func InternalKeyForAddr(wallet WalletController, netParams *chaincfg.Params,
 
 	walletAddr, err := wallet.AddressInfo(addr)
 	if err != nil {
+		// If the error is that the address can't be found, it is not
+		// an error, as callers can use the .Option() method to get an
+		// option value.
+		var managerErr waddrmgr.ManagerError
+		if errors.As(err, &managerErr) {
```
Comment (guggero) [inline lnwallet/interface.go:643]: nit: could fold into a single condition:

```
		if errors.As(err, &managerErr) &&
			managerErr.ErrorCode == waddrmgr.ErrAddressNotFound {

			return none, nil
		}
```

File: itest/lnd_coop_close_external_delivery_test.go [hunk 2]
```diff
@@ -10,13 +10,32 @@ import (
 )
 
 func testCoopCloseWithExternalDelivery(ht *lntest.HarnessTest) {
-	ht.Run("set delivery address at open", func(t *testing.T) {
+	ht.Run("set P2WPKH delivery address at open", func(t *testing.T) {
```
Comment (guggero) [inline itest/lnd_coop_close_external_delivery_test.go:13]: Nice, thanks for the additional tests!

**Review by guggero — APPROVED** [review]
Thanks for the fix! LGTM :tada:

File: lnwallet/interface.go [hunk 3]
```diff
@@ -636,6 +636,16 @@ func InternalKeyForAddr(wallet WalletController, netParams *chaincfg.Params,
 
 	walletAddr, err := wallet.AddressInfo(addr)
 	if err != nil {
+		// If the error is that the address can't be found, it is not
+		// an error, as callers can use the .Option() method to get an
```
Comment (yyforyongyu) [inline lnwallet/interface.go:640]: Not sure I follow here, could you elaborate a bit? If there's no internal key associated shouldn't we return an error?

File: itest/lnd_coop_close_external_delivery_test.go [hunk 4]
```diff
@@ -10,13 +10,32 @@ import (
...
 		tt := ht.Subtest(t)
-		testCoopCloseWithExternalDeliveryImpl(tt, true)
+		testCoopCloseWithExternalDeliveryImpl(
+			tt, true, lnrpc.AddressType_UNUSED_WITNESS_PUBKEY_HASH,
+		)
 	})
-	ht.Run("set delivery address at close", func(t *testing.T) {
+
+	ht.Run("set P2WPKH delivery address at close", func(t *testing.T) {
```
Comment (yyforyongyu) [inline itest/lnd_coop_close_external_delivery_test.go:20]: Nice tests - can we add `success := ht.Run` and stop the test if `!success`? this is helpful for future development as it's easier to debug that way.

File: lnwallet/interface.go (hunk 1 above)
Comment (starius) [reply lnwallet/interface.go]: Thanks! Fixed

File: lnwallet/interface.go (hunk 3 above)
Comment (starius) [reply lnwallet/interface.go]: I changed the comment:

```go
        walletAddr, err := wallet.AddressInfo(addr)
        if err != nil {
                // If the error is that the address can't be found, it is not
                // an error. This happens when any channel which is not a custom
                // taproot channel is cooperatively closed to an external P2TR
                // address. In this case there is no internal key associated
                // with the address. Callers can use the .Option() method to get
                // an option value.
```

File: itest/lnd_coop_close_external_delivery_test.go (hunk 4 above)
Comment (starius) [reply itest/lnd_coop_close_external_delivery_test.go]: Thanks! Done:

```go
        ok := ht.Run("set P2WPKH delivery address at open", func(t *testing.T) {
                tt := ht.Subtest(t)
                testCoopCloseWithExternalDeliveryImpl(
                        tt, true, lnrpc.AddressType_UNUSED_WITNESS_PUBKEY_HASH,
                )
        })
        // Abort the test if failed.
        if !ok {
                return
        }

...
```

File: itest/lnd_coop_close_external_delivery_test.go [hunk 5]
```diff
@@ -10,13 +10,44 @@ import (
...
+	ok := ht.Run("set P2WPKH delivery address at open", func(t *testing.T) {
 		tt := ht.Subtest(t)
-		testCoopCloseWithExternalDeliveryImpl(tt, true)
+		testCoopCloseWithExternalDeliveryImpl(
+			tt, true, lnrpc.AddressType_UNUSED_WITNESS_PUBKEY_HASH,
+		)
 	})
-	ht.Run("set delivery address at close", func(t *testing.T) {
+	// Abort the test if failed.
+	if !ok {
```
Comment (yyforyongyu) [inline itest/lnd_coop_close_external_delivery_test.go:20]: 🙏

File: lnwallet/interface.go [hunk 6]
```diff
@@ -636,6 +636,19 @@ func InternalKeyForAddr(wallet WalletController, netParams *chaincfg.Params,
 
 	walletAddr, err := wallet.AddressInfo(addr)
 	if err != nil {
+		// If the error is that the address can't be found, it is not
```
Comment (yyforyongyu) [inline lnwallet/interface.go:639]: 👍

File: lnwallet/interface.go [hunk 7]
```diff
@@ -636,6 +636,19 @@ func InternalKeyForAddr(wallet WalletController, netParams *chaincfg.Params,
...
+		// an error. This happens when any channel which is not a custom
+		// taproot channel is cooperatively closed to an external P2TR
+		// address. In this case there is no internal key associated
+		// with the address. Callers can use the .Option() method to get
```
Comment (yyforyongyu) [inline lnwallet/interface.go:643]: Q: how can the callers get the value, don't we return a `none` option here?

File: lnwallet/interface.go [hunk 8]
```diff
@@ -636,6 +636,19 @@ func InternalKeyForAddr(wallet WalletController, netParams *chaincfg.Params,
...
+		// an option value.
+		var managerErr waddrmgr.ManagerError
+		if errors.As(err, &managerErr) &&
+			managerErr.ErrorCode == waddrmgr.ErrAddressNotFound {
```
Comment (yyforyongyu) [inline lnwallet/interface.go:647]: I don't have the full pic here, but just wanna point out that there probably should have a dedicated method to do this job since this method is named `InternalKeyForAddr` and the caller would expect an error when there's no internal key. In addition, we are fixing the case "non-custom tap channel closed to an external addr", then what about "custom tap chan closed but no internal key found" case? Previously we'd have an error but now it's ignored. Is that handled elsewhere? cc @guggero

**Review by yyforyongyu — COMMENTED** [review]
A few more questions then this should be good to go!

File: lnwallet/interface.go (hunk 8 above)
Comment (guggero) [reply lnwallet/interface.go:647]: We only need the internal key in case of a custom (tap) channel. So for a normal taproot channel, we don't want this to error out and return a `None`. If we also return a `None` for custom (tap) channels, then that is fine _here_, as it will lead to an error elsewhere.
So the fix is good, we basically shouldn't error out for anything that can happen during normal operation (and specifying an _external_ shutdown address definitely is normal operation).

Basically the
```
if walletAddr == nil {
		return none, nil
}
```

further below is what we thought would happen if a key wasn't found. But apparently an error is returned, so we need this special case.

Comment (yyforyongyu) [reply lnwallet/interface.go:647]: cool thanks for the explanation!

**Review by yyforyongyu — APPROVED** [review]
LGTM 💃

File: lnwallet/interface.go (hunk 7 above)
Comment (starius) [reply lnwallet/interface.go:643]: I copy-pasted this part of the comment from the check below:

```go
        // No wallet addr. No error, but we'll return an nil error value here,
        // as callers can use the .Option() method to get an option value.
        if walletAddr == nil {
                return none, nil
        }
```

I propose to modify it in both cases to the following: "callers can inspect the returned option to determine if it has a value".

cc @guggero
